go/v1beta1/storage/filtering/filtering.go

```diff
@@ -141,59 +225,113 @@ func parseExpression(expression *expr.Expr) (*Query, error) {
 
+	target, err := assertString(parsedTarget)
+	if err != nil {
+		return nil, err
+	}
+
+	arg, err := assertString(parsedArg)
+	if err != nil {
+		return nil, err
+	}
+
+	switch callExpr.Function {
+	case overloads.StartsWith:
 		return &Query{
 			Prefix: &Term{
-				leftTerm: rightTerm,
+				target: arg,
 			},
 		}, nil
 	case overloads.Contains:
-		leftTerm, rightTerm, err := getSimpleExpressionTerms(leftArg, rightArg)
-		if err != nil {
-			return nil, err
-		}
-
-		// special characters need to be escaped via "\"
-		query := fmt.Sprintf("*%s*", elasticsearchSpecialCharacterRegex.ReplaceAllString(rightTerm, `\$1`))
-
 		return &Query{
 			QueryString: &QueryString{
-				DefaultField: leftTerm,
-				Query:        query,
+				DefaultField: target,
+				Query:        fmt.Sprintf("*%s*", elasticsearchSpecialCharacterRegex.ReplaceAllString(arg, `\$1`)),
 			},
 		}, nil
-	default:
-		return nil, fmt.Errorf("unknown parse expression function: %s", function)
 	}
+
+	return nil, fmt.Errorf("unrecognized function: %s", callExpr.Function)
 }
 
-// converts left and right call expressions into simple term strings.
-// this function should be used at the top of the `parseExpression` call stack.
-func getSimpleExpressionTerms(leftArg, rightArg *expr.Expr) (leftTerm, rightTerm string, err error) {
-	if _, ok := leftArg.GetExprKind().(*expr.Expr_IdentExpr); ok {
-		leftTerm = leftArg.GetIdentExpr().Name
+func (f *filterer) visitNestedFilterCall(expression *expr.Expr, depth string) (interface{}, error) {
+	callExpr := expression.GetCallExpr()
+	targetExpr := callExpr.Target
+
+	parsedTarget, err := f.visit(targetExpr, depth)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(callExpr.Args) != 1 {
+		return nil, fmt.Errorf("invalid number of arguments")
 	}
 
-	if _, ok := leftArg.GetExprKind().(*expr.Expr_ConstExpr); ok {
-		leftTerm = leftArg.GetConstExpr().GetStringValue()
+	argExpr := callExpr.Args[0]
+	target, err := assertString(parsedTarget)
+	if err != nil {
+		return nil, err
 	}
 
-	if _, ok := rightArg.GetExprKind().(*expr.Expr_IdentExpr); ok {
-		rightTerm = rightArg.GetIdentExpr().Name
+	newDepth := target
+	if depth != "" {
+		newDepth = fmt.Sprintf("%s.%s", depth, newDepth)
 	}
 
-	if _, ok := rightArg.GetExprKind().(*expr.Expr_ConstExpr); ok {
-		rightTerm = rightArg.GetConstExpr().GetStringValue()
+	maybeNestedQuery, err := f.visit(argExpr, newDepth)
+	if err != nil {
+		return nil, err
+	}
+	nestedQuery, ok := maybeNestedQuery.(*Query)
+	if !ok {
+		return nil, fmt.Errorf("nested expression was not a valid query")
 	}
 
-	if leftTerm == "" || rightTerm == "" {
-		err = fmt.Errorf("encountered unexpected expression kinds when evaluating filter: %T, %T", leftArg.GetExprKind(), rightArg.GetExprKind())
+	return &Query{
+		Nested: &Nested{
+			Path:  target,
+			Query: nestedQuery,
+		},
+	}, nil
+}
+
+func assertString(value interface{}) (string, error) {
+	stringValue, ok := value.(string)
+	if !ok {
+		return "", fmt.Errorf("not a string")
```

Comment:
It seems like there are a lot of paths where an expression could fail to parse and the only feedback would be "not a string". Does this end up being returned by the API and if so can we improve the messaging?

Reply:
Added a little bit more information to the error so that the final error from the API would read "error while parsing filter expression: expected 1 to have type string but was int64" for a query like `resource.uri == 1`. 

In main, that same query comes back as "error while parsing filter expression: encountered unexpected expression kinds when evaluating filter: *expr.Expr_ConstExpr, *expr.Expr_ConstExpr".